doc/cascadia/Command-Palette-spec.md

```diff
@@ -0,0 +1,269 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2019-08-01
+last updated: 2019-08-01
+issue id: 2046
+---
+
+# Spec Title
+
+## Abstract
+
+This spec covers the addition of a "command palette" to the Windows Terminal.
+The Command Palette is a GUI that the user can activate to search for and
+execute commands. Beneficially, the command palette allows the user to execute
+commands _even if they aren't bound to a keybinding_.
+
+## Inspiration
+
+This feature is largely inspired by the "Command Palette" in text editors like
+VsCode, Sublime Text and others.
+
+This spec was initially drafted in [a
+comment](https://github.com/microsoft/terminal/issues/2046#issuecomment-514219791)
+in [#2046]. That was authored during the annual Microsoft Hackathon, where I
+proceeded to prototype the solution. This spec is influenced by things I learned
+prototyping.
+
+## Solution Design
+
+First off, for the sake of clarity, we'll rename the `command` of a keybinding
+to `action`. This will help keep the mental model between commands and actions
+clearer. When deserializing keybindings, we'll include a check for the old
+`command` key to migrate it.
+
+We'll introduce a new top-level array to the user settings, under the key
+`commands`. `commands` will contain an array of commands, each with the
+following schema:
+
+```js
+{
+    "name": string,
+    "action": string,
+    "iconPath": string
+}
+```
+
+Command names should be human-friendly names of actions, though they don't need
+to necessarily be related to the action that it fires. For example, a command
+with `newTab` as the action could have `"Open New Tab"` as the name.
+
+The command will be parsed into a new class, `Command`:
+```c++
+class Command
+{
+    winrt::hstring Name();
+    winrt::TerminalApp::ShortcutAction Action();
+    winrt::hstring IconPath();
+}
+```
+We'll add another structure in GlobalAppSettings to hold all these actions. It

```

Comment:
/me: put a newline here to not have GH's markdown syntax highlighter beef it here.